services/src/main/java/io/scalecube/services/Microservices.java

```diff
@@ -104,85 +107,210 @@
+            null,
+            thread -> log.warn("{} failed on doShutdown(): {}", this, thread.toString()),
+            () -> log.debug("Shutdown {}", this));
+  }
+
+  public Microservices(Microservices msBase) {
+    this();
+    this.tags = new HashMap<>(msBase.tags);
+    this.serviceInfos = new ArrayList<>(msBase.serviceInfos);
+    this.serviceProviders = new ArrayList<>(msBase.serviceProviders);
+    this.serviceRegistry = msBase.serviceRegistry;
+    this.methodRegistry = msBase.methodRegistry;
+    this.gatewayBootstrap = msBase.gatewayBootstrap;
+    this.discovery = msBase.discovery;
+    this.errorMapper = msBase.errorMapper;
+    this.transportBootstrap = msBase.transportBootstrap;
+    this.discoveryOptions = msBase.discoveryOptions;
+    this.metrics = msBase.metrics;
+  }
+
+  public Microservices metrics(Metrics metrics) {
+    Microservices msNew = new Microservices(this);
+    msNew.metrics = metrics;
+    return msNew;
+  }
+
+  public Microservices metrics(MetricRegistry metrics) {
+    Microservices msNew = new Microservices(this);
+    msNew.metrics = new Metrics(metrics);
+    return msNew;
+  }
+
+  public Microservices tags(Map<String, String> tags) {
+    Microservices msNew = new Microservices(this);
+    msNew.tags.putAll(tags);
+    return msNew;
+  }
+
+  public Microservices services(ServiceProvider serviceProvider) {
+    Microservices msNew = new Microservices(this);
+    msNew.serviceProviders.add(serviceProvider);
+    return msNew;
+  }
+
+  public Microservices services(Object... services) {
+    Microservices msNew = new Microservices(this);
+    msNew.serviceProviders.add(
+        call ->
+            Arrays.stream(services)
+                .map(
+                    s ->
+                        s instanceof ServiceInfo
+                            ? (ServiceInfo) s
+                            : ServiceInfo.fromServiceInstance(s).build())
+                .collect(Collectors.toList()));
+    return msNew;
+  }
+
+  public Microservices services(ServiceInfo... services) {
+    Microservices msNew = new Microservices(this);
+    msNew.serviceProviders.add(call -> Arrays.stream(services).collect(Collectors.toList()));
+    return msNew;
+  }
+
+  public Microservices serviceRegistry(ServiceRegistry serviceRegistry) {
+    Microservices msNew = new Microservices(this);
+    msNew.serviceRegistry = serviceRegistry;
+    return msNew;
+  }
+
+  public Microservices methodRegistry(ServiceMethodRegistry methodRegistry) {
+    Microservices msNew = new Microservices(this);
+    msNew.methodRegistry = methodRegistry;
+    return msNew;
+  }
+
+  public Microservices discovery(ServiceDiscovery discovery) {
+    Microservices msNew = new Microservices(this);
+    msNew.discovery = discovery;
+    return msNew;
+  }
+
+  public Microservices discovery(Consumer<ServiceDiscoveryConfig.Builder> discoveryOptions) {
+    Microservices msNew = new Microservices(this);
+    msNew.discoveryOptions = discoveryOptions;
```

Comment:
I think it looks wrong a bit. I'd like to suppose the following solution for it:
```java
    if (msNew.discoveryOptions != null) {
      msNew.discoveryOptions = msNew.discoveryOptions.andThen(discoveryOptions);
    } else {
      msNew.discoveryOptions = discoveryOptions;
    }
```
It will give us  an opportunity do the next one:
```java
        Microservices.builder()
            .discovery(options -> options.seeds(seed.discovery().address()))
            .discovery(options -> options.port(123))
            .discovery(options -> options.memberHost("member"))
```

Reply:
@ronenhamias do you mind of null value or compose of two consumers?
so now, the null value of this consumer is legal state. For example when we create a seed it doesn't have discovery option consumer at all.

Reply:
i think that in either case it should not be null.

Reply:
@segabriel @ronenhamias 
Things what you are discussing here is not a part of this pull request.

This pull request is about _refactoring_ which generally means _change code structure without changing behavior_.
@segabriel your proposition https://github.com/scalecube/scalecube-services/pull/511#discussion_r252613337 should be discussed out of this pull request as separate issue, because in previous implementation (with Builder) we had totally the same behavior regarding null-checking.